include/nd_memory.h

```diff
@@ -0,0 +1,122 @@
+	};
+
+	size_t offset_in_source = 0; ///< offset in the source allocation, in bytes, of the first chunk to copy.
+	size_t offset_in_dest = 0;   ///< offset in the destination allocation, in bytes, of the first chunk to copy.
+	int num_complex_strides = 0; ///< number of strides which are not contiguous in either source or destination.
+	stride_dimension strides[2]; ///< in the 3D / 2-complex case, strides[0] is the outer and strides[1] the inner stride.
+	size_t contiguous_size = 0;  ///< number of contiguous bytes in the last dimension.
+
+	friend bool operator==(const nd_copy_layout& lhs, const nd_copy_layout& rhs) {
+		return lhs.offset_in_source == rhs.offset_in_source && lhs.offset_in_dest == rhs.offset_in_dest && lhs.contiguous_size == rhs.contiguous_size
+		       && lhs.num_complex_strides == rhs.num_complex_strides && lhs.strides[0] == rhs.strides[0] && lhs.strides[1] == rhs.strides[1];
+	}
+
+	friend bool operator!=(const nd_copy_layout& lhs, const nd_copy_layout& rhs) { return !(lhs == rhs); }
+};
+
+/// Computes the minimum number of complex strides required to describe a box-shaped copy between box-shaped allocations.
+inline nd_copy_layout layout_nd_copy(const range<3>& source_range, const range<3>& dest_range, const id<3>& offset_in_source, const id<3>& offset_in_dest,
+    const range<3>& copy_range, size_t elem_size) {
+	assert(all_true(offset_in_source + copy_range <= source_range));
+	assert(all_true(offset_in_dest + copy_range <= dest_range));
+
+	if(copy_range.size() == 0) return {};
+
+	nd_copy_layout layout;
+	layout.offset_in_source = get_linear_index(source_range, offset_in_source) * elem_size;
+	layout.offset_in_dest = get_linear_index(dest_range, offset_in_dest) * elem_size;
+	layout.contiguous_size = elem_size;
+	size_t* current_range = &layout.contiguous_size; // we first maximize the contiguous range, then the range of the 0th / 1st stride.
+	size_t next_source_stride = elem_size;
+	size_t next_dest_stride = elem_size;
+	bool contiguous = true; // when false, the next-higher non-trivial dimension must insert a stride
+	for(int d = 2; d >= 0; --d) {
+		if(!contiguous && copy_range[d] != 1) { // if range is 1, we can postpone (and possibly avoid) inserting a stride
+			++layout.num_complex_strides;
+			layout.strides[1] = layout.strides[0];
+			layout.strides[0] = {next_source_stride, next_dest_stride, 1};
+			current_range = &layout.strides[0].count;
+			contiguous = true;
+		}
+		next_source_stride *= source_range[d];
+		next_dest_stride *= dest_range[d];
+		*current_range *= copy_range[d];
+		if(source_range[d] != copy_range[d] || dest_range[d] != copy_range[d]) { contiguous = false; }
+	}
+
+	return layout;
+}
+
+/// For every contiguous chunk in an nd-copy, invoke `f(byte_offset_in_source, byte_offset_in_dest, chunk_bytes)`.
+template <typename F>
+inline void for_each_contiguous_chunk(const nd_copy_layout& layout, F&& f) {
```

Comment:
I would rename this to `for_each_contiguous_chunk_do` or `for_each_contiguous_chunk_apply`, but this is just a preference.

Reply:
Prior art: `host_utils::for_each_item()`, `region_map::impl::for_each()`, and `std::for_each()`.